### PR TITLE
feature: fetch post endpoint

### PR DESCRIPTION
**Pull request by ryanycw**
## Summary

In this pull request, we built the API to fetch posts from the database. Additionally, we refactored the synchronizer implementation to ensure that posts stored in the database have synchronized status with on-chain transactions.

## Details

The critical changes would be in the following files:
- packages/relay/src/synchornizer.ts
- packages/relay/src/routes/post.ts
- packages/relay/src/singletons/schema.ts

After sending a transaction, we store it in our DB. Initially, the post status is 0, indicating it was not found on-chain. Once the synchronizer locates the post on-chain, the status updates to 1.

## Verification Steps

The following image is an example of testing under local environment with Postman.

### Posting

image
```
{
    "content": "12312",
    "publicSignals": [
        "11114933767258551142419152624194302090535444181093381511331956500611967564118",
        "3988878084682830050978437313015151192452737299514626821739359355499244837920",
        "4451353010164519564206090957593902219217783864052023977979736464818176",
        "0"
    ],
    "proof": [
        "13811368394148964750623122474289487870144078130603099319470943530205297015580",
        "19347858422244784665570916184808112825340671648025872330141085455269033702079",
        "10633064249727374883942777964008590711107517583764228847996045637733089922795",
        "1880440864244341948858931363793603145441123533718128059219919698049591683330",
        "14356538636969954241468117209982959029790216139263827811083879468062630522423",
        "19706076580715843084508065049092032826404988347207421043942067334527164172863",
        "9141826036467092630741616376559048604989630069085247882180407783557429414647",
        "435899010385575362246012338112682113841445828034870524516369387190370801050"
    ]
}
```
### Fetching Posts

image

## References
- [@unirep/core 2.0.0-beta-3](https://github.com/Unirep/Unirep/blob/v2.0.0-beta-3/packages/core/src/Synchronizer.ts#L151)
- [Unirep Social with @unirep/core 2.0.0-beta-3](https://github.com/Unirep/Unirep-Social/blob/2e1b3376399ec65f64e77fe81fef6c0227a5105e/packages/backend/src/Synchronizer.ts#L200)
- [@unirep/core 1.0.4](https://github.com/Unirep/Unirep/blob/v1.0.4/packages/core/src/Synchronizer.ts#L199)
- [Unirep Social with @unirep/core 1.0.4](https://github.com/Unirep/Unirep-Social/blob/main/packages/core/src/synchronizer.ts#L83)

### CODE REVIEW

File: packages/relay/src/synchornizer.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        return {
+            ...super.contracts,
+            [tempUnirepSocialContract.address]: {
+                contract: tempUnirepSocialContract,
+                eventNames: ['Post'],
+            },
+        }
+    }
+
+    async handlePost({ event, db, decodedData }: EventHandlerArgs) {
```
[the remark this answers is not in the document]
Comment (ryanycw) [reply packages/relay/src/synchornizer.ts:46]: 這個會在 Synchronizer 裡被使用，詳細可以參考 References 的前兩個連結，主要就是會去抓 handle<SmartContractEvent> function 來處理 event，如果沒有這個 function 就會變成 unrecognized event

File: packages/relay/src/singletons/schema.ts [hunk 2]
```diff
@@ -20,6 +20,42 @@ const _schema = [
...
+                name: 'downCount',
+                type: 'Int',
+                default: () => 0,
+            },
+            {
+                name: 'voteSum',
+                type: 'Int',
+                default: () => 0,
+            },
+            ['status', 'Int'],
```
[the remark this answers is not in the document]
Comment (ryanycw) [reply packages/relay/src/singletons/schema.ts:51]: 收到！

File: packages/relay/src/synchornizer.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        } else {
+            db.create('Post', {
+                hashedContent,
+                onChainId,
+                epochKey,
+                epoch,
+                transactionHash,
+                status: 1,
+            })
+        }
```
[the remark this answers is not in the document]
Comment (ryanycw) [reply packages/relay/src/synchornizer.ts:79]: 收到！

File: packages/relay/src/singletons/schema.ts [hunk 4]
```diff
@@ -20,6 +20,42 @@ const _schema = [
...
         ],
     },
+    {
+        name: 'Post',
+        rows: [
+            {
+                name: 'publishedAt',
+                type: 'Int',
+                default: () => +new Date(),
+            },
```
[the remark this answers is not in the document]
Comment (ryanycw) [reply packages/relay/src/singletons/schema.ts:30]: 我原本的想法是如果是上鏈時間應該會有許多 Post 在同一時間，存 relayer 收到的時間，應該會更貼近使用者發文的時間

File: packages/relay/src/synchornizer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xiawpohr) [reply packages/relay/src/synchornizer.ts:46]: 了解

**Comment by ryanycw** [comment]
@Xiawpohr Please let me know if any other issues should be addressed.

**Comment by Xiawpohr** [comment]
I think there is no more issues.

**Pull request merged**